graphics.c

```diff
@@ -0,0 +1,158 @@
+#include "graphics.h"
+#include "display.h"
+
+static inline void vGraphicsPixelWrite(color pixelColor)
+{
+    vDisplayPixelWrite(pixelColor.colorSeperate.redValue, pixelColor.colorSeperate.greenValue, pixelColor.colorSeperate.blueValue);
+}
+
+
+curserStruct curserA,curserB;
+
+void graphicsInit()
+{
+
+vDisplayClear();    //blackout the display
+
+curserA.curserColor=ACURSERCOLOR;
+curserB.curserColor=BCURSERCOLOR;
+
+printGrid();    //print the grid
+
+}
+
+//printRectangle prints a Rectangle
+
+void printRectangle(uint16_t xStart,uint16_t xEnd,uint16_t yStart,uint16_t yEnd,color lineColor)
+{
+    uint16_t row,column;
+
+    vDisplayWindowSet(xStart,xEnd,yStart,yEnd);
+    vDisplayStartPixelWrite();
+
+    for(row=yStart;row<=yEnd;row++)
+    {
+        for(column=xStart;column<=xEnd;column++)
+        {
+            vGraphicsPixelWrite(lineColor);
+        }
+    }
+}
+
+
+//printLineBresenham prints a line between two points, including the second point, excluding the first point
+void printLineBresenham(uint16_t xA,uint16_t yA,uint16_t xB,uint16_t yB,color lineColor)
+{
+uint16_t x=xA,y=yA,dX=xB-xA,dY;
+int16_t delta,step=1;
+uint8_t risingLine=1;
+
+if(yB-yA > 0)   //decision if line shall be printed in negative or positive y-direction
+{
+    dY=yB-yA;
+}
+else
+{
+    dY=yA-yB;
+    risingLine=0;
+    step=-1;
+}
+
+
+delta=dY>>1;
+
+while(y!=yB)
+{
+    delta-=dX;
+
+    y+=step;
+
+    if(delta<=0)
+    {
+        delta+=dY;
+        x++;
+    }
+
+    vDisplayWindowSet(x,x,271-y,271-y);
+    vDisplayStartPixelWrite();
+    vGraphicsPixelWrite(lineColor);
+}
+
+}
+
+/* printWave prints the wave (data saved in printBuffer) from xStart to xEnd. In cases where it is needed it uses the bresenHam algorithm to print lines.
+ * PREVIOUS WAVES NEED TO BE DELETED MANUALY*/
+
+void printWave(uint8_t* printBuffer,uint16_t xStart,uint16_t xEnd,color waveColor)
```

Comment:
- In UI-Modul verschieben
- `vGraphicsDrawWave`
- Messdaten werden als `int16_t` dargestellt -> `int16_t* p_aiSampleBuffer`
- Sample buffer ist immer 480 Werte lang (hierfür Makro aus sampler.h verwenden)
- Start- und Endpunkt entfernen. Hierfür jeden Sample mit `SAMPLER_SAMPLE_INVALID` abgleichen. Ist das Sample ungültig, soll es nicht gezeichnet werden.

Bleibt also `vGraphicsDrawWave(tsColor sWaveColor)` übrig

Reply:
printWave ist so gedacht, dass verschieden lange Abschnitte der Samples gezeichnet werden können. Das hat verschiedene Anwendungsbereiche. Zum Beispiel zum erneuern der Welle wenn der Curser neu gezeichnet wird. Man kann die Funktion auch benutzen, um bei hoher timebase einzelne Messwerte zu zeichnen. Sie ist also universell einsetzbar durch die Übergabe eines Start- und Endwertes. Der Buffer selber ist immer 480 Elemente lang. printBuffer ist der Buffer für die Punkte die gezeichnet werden sollen, also uint8_t*.

Reply:
Diese Beschreibung gerne so in die detaillierte Beschreibung im Kommentar tun.
Der Messwerte-Speicher vom Sampler ist mit `SAMPLER_SAMPLE_INVALID` (=-1) initialisiert. Der Wert stellt also die Stellen dar, wo noch kein Messwert eingetragen wurde.

Reply:
Den "printBuffer" am Besten modulintern verwalten, damit man sich außerhalb vom Graphics-Modul beim Prozeduraufruf keine Gedanken mehr darum machen muss.